### ISSUE TITLE
Body as Input Parameter

### ISSUE THREAD

**Issue by billjwolff** [Azure/azure-functions-sql-extension#302]
I want to pass a large string to a stored procedure. That won't fit as a query parameter but will work as the request body. Is there a way to set the parameter before the input bindings executes? I can pass the string (JSON object) in the body, so I want to do something like this:

  context.bindings.JSONInput = JSON.stringify(req.body);

**Comment by Charles-Gagnon** [comment]
I don't think this is possible from what I'm seeing about the binding expressions - @dzsquared do you know of a way to do this?

You should be able to use the SqlCommand parameter and then modify the query with the POST data though before running the command. See https://github.com/Azure/azure-functions-sql-extension/blob/main/samples/samples-csharp/InputBindingSamples/GetProductsSqlCommand.cs for example usage.

**Comment by dzsquared** [comment]
The good news is that it can be done by handling the HTTP trigger with a [custom object](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=in-process%2Cfunctionsv2&pivots=programming-language-csharp#payload).
> For a custom type, the runtime tries to parse the JSON request body to set the object properties.

```c#
    [FunctionName("HttpBodyToProc")]
        public static IActionResult Run(
            [HttpTrigger(AuthorizationLevel.Function, "post", Route = null)] HttpBodyObj reqbody,
            [Sql("dbo.testInput",
            CommandType = System.Data.CommandType.StoredProcedure,
            Parameters = "@param1={message}",
            ConnectionStringSetting = "SqlConnectionString")] IEnumerable<Object> result,
            ILogger log)
        {
            log.LogInformation("C# HTTP trigger with SQL Input Binding function processed a request.");

            return new OkObjectResult(result);
        }
    }

    // body object to parse JSON to POCO
    public class HttpBodyObj
    {
        public string message { get; set; }
    }
```


would take this POST body
```
{
  "message": "Hello World"
}
```
and pass "Hello World" to the stored proc


However, this gives the input bindings access to the properties of the object - where you want to hand it a bunch of JSON.

I was able to prototype this by wrapping my JSON in more JSON 🙃 and defining the *set* accessor for the outer property to deserialize the inner JSON to string.

```c#
[FunctionName("HttpBodyToProc")]
        public static IActionResult Run(
            [HttpTrigger(AuthorizationLevel.Function, "post", Route = null)] HttpJsonBodyObj reqbody,
            [Sql("dbo.testInput",
            CommandType = System.Data.CommandType.StoredProcedure,
            Parameters = "@param1={bodyjson}",
            ConnectionStringSetting = "SqlConnectionString")] IEnumerable<Object> result,
            ILogger log)
        {
            log.LogInformation("C# HTTP trigger with SQL Input Binding function processed a request.");

            return new OkObjectResult(result);
        }
    }

    // body object to parse JSON to nested JSON
    public class HttpJsonBodyObj
    {
        // define bodyjson set as convert nested JSON to string
        private string _bodyjson;
        public dynamic bodyjson { 
            get => _bodyjson;
            set {
                // jobject to string
                _bodyjson = Newtonsoft.Json.JsonConvert.SerializeObject(value);
                //value;
            }
        }
    }
```


would take this POST body
```
{
  "bodyjson": {
    "message": "Hello World",
    "message2": "Hello World2",
    "message3": "Hello World3"
  }
}
```
and pass 
```
{
    "message": "Hello World",
    "message2": "Hello World2",
    "message3": "Hello World3"
  }
```
 to the stored proc




So to recap, instead of using the `HttpRequest req` output of the trigger, you define a custom object (such as `HttpJsonBodyObj reqbody` above) and specify its properties as the parameter values to your stored proc.

**Comment by Charles-Gagnon** [comment]
@billjwolff Drew and I looked at this some more and while his first example works (mostly) fine the second one unfortunately isn't going to work because of how we do parameter parsing

https://github.com/Azure/azure-functions-sql-extension/blob/main/src/SqlBindingUtilities.cs#L68

Because of the [AutoResolve] attribute on the Parameters property that can only be a string, which we then parse out into the separate parameters to pass to the function. This means that the names and values are unable to have , or = in them or the parsing will break. 

Keep in mind this will also mean that the first solution will break if the string may have , or = in it as well. 

So if that isn't a concern then you should be fine to use that approach. But if you need the second option or will have ,/='s in your parameters then we'll have to find some other workaround. One approach you could consider is to encode the field before it gets to our binding : 

```csharp
    public class HttpEncodedBodyObj
    {
        private string _encodedMessage;
        public dynamic message
        {
            get => _encodedMessage;
            set
            {
                _encodedMessage = value.Replace(",", "%2C");
            }
        }
    }
```

and then decode it in your procedure. 

I'm going to close this issue since the original question as asked is technically supported, and we already have https://github.com/Azure/azure-functions-sql-extension/issues/6 for tracking the parameter parsing issues. Let us know if you have any further information or questions!

**Comment by Charles-Gagnon** [comment]
You can't use the full body directly, but this doc shows some examples of using values passed in the body 

https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-expressions-patterns#json-payloads

**Comment by mumurug-MSFT** [comment]
@Charles-Gagnon Thanks, it helps.
